### PR TITLE
ENH: Add Limited memory BFGS, closes #621

### LINKED ISSUE #621: limited memory BFGS
We should implement the [Limited-memory BFGS](https://en.wikipedia.org/wiki/Limited-memory_BFGS) algorithm.

### PR DESCRIPTION
**Pull request by adler-j**
Added memory limited option to BFGS solver.

This is a merge into the #569 branch since it needs functionality from there.

Edit:
Added more functionality:

* Fixed legacy issues including a bug with the derivative of `ConstantOperator`. `ZeroOperator` is now a free standing operator.
* Added LBFGS example.

### CODE REVIEW

File: odl/operator/default_ops.py [hunk 1]
```diff
@@ -834,99 +816,71 @@ def __str__(self):
...
-    def vector(self):
-        """The constant operator range element to subtract."""
-        return self.__vector
+        super().__init__(domain, range, linear=True)
 
     def _call(self, x, out=None):
-        """Evaluate the residual at ``x`` and write to ``out`` if given."""
-        if out is None:
-            out = self.operator(x)
+        """Return the constant vector or assign it to ``out``."""
```
Comment (aringh) [inline odl/operator/default_ops.py]: Explicit would be to say `zero vector` and not `constant vector`.

File: odl/operator/operator.py [hunk 2]
```diff
@@ -1158,6 +1155,97 @@ def __str__(self):
...
+        super().__init__(operator.domain, operator.range)
+
+    @property
+    def operator(self):
+        """The operator to apply."""
+        return self.__operator
+
+    @property
+    def vector(self):
+        """The constant operator range element to subtract."""
```
Comment (aringh) [inline odl/operator/operator.py]: In the definition above and the `_call` below it looks like the vector is added.

File: odl/operator/operator.py [hunk 3]
```diff
@@ -1158,6 +1155,97 @@ def __str__(self):
...
+        ----------
+        point : `domain` element
+            Any element in the domain where the derivative should be taken
+
+        Examples
+        --------
+        >>> import odl
+        >>> r3 = odl.rn(3)
+        >>> op = odl.IdentityOperator(r3)
+        >>> res = odl.OperatorVectorSum(op, r3.element([1, 2, 3]))
```
Comment (aringh) [inline odl/operator/operator.py]: Very minor: I guess `res` is with "residual" in mind. However this operator is the sum.

File: odl/solvers/smooth/newton.py [hunk 4]
```diff
@@ -34,6 +34,54 @@
...
+    Parameters
+    ----------
+    s : `sequence` of `LinearSpaceElement`
+        The ``s`` coefficients in the BFGS update, see Notes.
+    y : `sequence` of `LinearSpaceElement`
+        The ``y`` coefficients in the BFGS update, see Notes.
+    x : `LinearSpaceElement`
+        Point in which to evaluate the product.
+    hessinv_estimate : `Operator`, optional
+        Initial estimate of the hessian ``H0^-1``.
```
Comment (aringh) [inline odl/solvers/smooth/newton.py]: What happens if `hessinv_estimate` is not give? Taken as identity?

Edit: I guess this is only called from the bfgs method, and there it says that identity operator is default. Maybe that is enough...

**Comment by adler-j** [comment]
Merged this into the other big branch (after fixing @aringh comments) review can be done over there instead.

**Pull request merged**